nano/node/node.cpp

```diff
@@ -3748,6 +3750,89 @@ std::unique_ptr<seq_con_info_component> collect_seq_con_info (active_transaction
 	return composite;
 }
 }
+
+/**
+ * Confirms the height of the block for the block hash passed in. If this has implicitly
+ * confirms other blocks then check if these are open/receive blocks, and if so follow
+ * the source of these blocks iteratively confirm them as well.
+ */
+void nano::node::add_confirmation_heights (nano::transaction const & transaction, nano::block_hash const & hash)
+{
+	std::stack<nano::block_hash, std::vector<nano::block_hash>> open_receive_blocks;
+	auto current = hash;
+
+	do
+	{
+		if (!open_receive_blocks.empty ())
+		{
+			current = open_receive_blocks.top ();
+			open_receive_blocks.pop ();
+
+			auto block (store.block_get (transaction, current));
+			if (block != nullptr)
+			{
+				nano::block_hash source_hash = block->source ();
```

Comment:
Probably it have to be a bit more complex. We don't have to check block existence if source_hash .is_zero ()
For state blocks need to check link () instead if link is not 0, not epoch link & that current state block !is_send ()
Instead of check block_get + (source_block != nullptr) can use source_exists (...) function

Reply:
Update: considering balance check for state blocks below expensive !is_send () not required

Actually why not merge this 2 parts? `if (!open_receive_blocks.empty ()) { ... }` & `// First check legacy receive/open`
I guess it can save several iops. If put to stack source () for receive/open blocks & link () for state blocks after balance check `if (state->hashables.balance > ledger.balance (transaction, previous))`

Reply:
Think I understand, please check to confirm though, thanks :)

Reply:
Seems great!